core/src/client/mod.rs

```diff
@@ -592,9 +682,233 @@ impl<'a, R> IntoIterator for BatchResponse<'a, R> {
+					self.drain();
+					return Err(SubscriptionError::TooSlow);
+				}
+			}
+			InnerStrategy::DropLatest(full) => {
+				let dropped = {
+					let mut lock = full.lock().unwrap();
+					let dropped = *lock;
+					*lock = false;
+					dropped
+				};
+
+				if dropped {
+					return Err(SubscriptionError::TooSlow);
+				}
+			}
+			InnerStrategy::Close | InnerStrategy::DropOldest => (),
+		};
+
+		Ok(())
+	}
+
+	fn drain(&mut self) {
+		self.inner = self.inner.new_receiver();
+	}
+}
+
+impl Stream for SubscriptionRx {
+	type Item = Result<serde_json::Value, SubscriptionError>;
+
+	fn poll_next(
+		mut self: std::pin::Pin<&mut Self>,
+		cx: &mut std::task::Context<'_>,
+	) -> std::task::Poll<Option<Self::Item>> {
+		if let Err(e) = self.read_shared_state() {
+			return Poll::Ready(Some(Err(e)));
+		}
+
+		let this = Pin::new(&mut self.inner);
+
+		let res = match futures_util::ready!(this.poll_recv(cx)) {
+			Some(Ok(v)) => Some(Ok(v)),
+			Some(Err(RecvError::Closed)) => None,
+			Some(Err(RecvError::Overflowed(_))) => Some(Err(SubscriptionError::TooSlow)),
+			None => None,
+		};
+
+		Poll::Ready(res)
+	}
+}
+
+pub(crate) fn subscription_stream(config: SubscriptionConfig) -> (SubscriptionTx, SubscriptionRx) {
+	let (mut tx, rx) = async_broadcast::broadcast(config.max_capacity);
+
+	if let LaggingStrategy::DropOldest = config.strategy {
+		tx.set_overflow(true);
+	}
+
+	let shared_strategy: InnerStrategy = config.strategy.into();
+
+	(
+		SubscriptionTx { inner: tx, strategy: shared_strategy.clone() },
+		SubscriptionRx { inner: rx, strategy: shared_strategy },
+	)
+}
+
+/// Individual subscription configuration.
+///
+/// Subscriptions are implemented on top of a bounded channel which may suffer from
+/// slow receiver problem" where it's an item can't fit in the buffer.
+///
+/// By this configuration it's possible to configure such scearios such
+/// how big the max buffer capacity, whether it's okay to drop item and similar.
+///
+/// See [`LaggingStrategy`] for more information about different strategies to use.
+#[derive(Debug, Copy, Clone)]
+pub struct SubscriptionConfig {
+	max_capacity: usize,
+	strategy: LaggingStrategy,
+}
+
+impl Default for SubscriptionConfig {
+	fn default() -> Self {
+		Self { max_capacity: 16, strategy: LaggingStrategy::Close }
+	}
+}
+
+impl SubscriptionConfig {
+	/// Create new [`SubscriptionConfig`]
+	pub fn new() -> Self {
+		Self::default()
+	}
+
+	/// Set the capacity for the subscription, this will allocate buffer with the capacity.
+	///
+	/// # Panics
+	///
+	/// This method panics if `capacity == 0`
+	pub fn set_capacity(mut self, max_cap: usize) -> Self {
+		self.max_capacity = max_cap;
+		self
+	}
+
+	/// The subscription may suffer from the slow-receiver problem
+	pub fn set_lagging_strategy(mut self, strategy: LaggingStrategy) -> Self {
+		self.strategy = strategy;
+		self
+	}
+}
+
+/// Strategy to use when a message could not be handled
+#[derive(Debug, Copy, Clone)]
+pub enum LaggingStrategy {
```

Comment:
dq: How easy would it be for a user to define their own custom strategy?

Reply:
it's not possible because we control stream/recv impl in that sense what James suggested is probably nicer and more generic...

but this should cover 99% of the use cases I reckon